core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,387 @@
+      final PendingTransaction lockOwner = waitForTransactionsToComplete(pendingTransactions, expectedEndTime);
+
+      // Then try to acquire a lock
+      if (trace) {
+         log.tracef("Finished waiting for other potential lockers. Timed-Out? %b", lockOwner != null);
+      }
+
+      if (lockOwner != null) {
+         timeout(lockOwner, globalTransaction);
+      }
+
+      return timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS);
+   }
+
+   private void timeout(PendingTransaction lockOwner, GlobalTransaction thisGlobalTransaction) {
+      throw new TimeoutException(format("Could not acquire lock on %s in behalf of transaction %s. Current owner %s.",
+                                        lockOwner.commonKey, thisGlobalTransaction,
+                                        lockOwner.cacheTransaction.getGlobalTransaction()));
+   }
+
+   private PendingTransaction waitForTransactionsToComplete(Collection<PendingTransaction> transactionsToCheck,
+                                                            long expectedEndTime) throws InterruptedException {
+      if (transactionsToCheck.isEmpty()) {
+         return null;
+      }
+      for (PendingTransaction tx : transactionsToCheck) {
+         long remaining;
+         if ((remaining = timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS)) > 0) {
+            if (!tx.cacheTransaction.waitForLockRelease(remaining)) {
+               return tx;
+            }
+         }
+      }
+      return null;
+   }
+
+   private Collection<PendingTransaction> getTransactionWithLockedKey(int transactionTopologyId,
+                                                                      Object key,
+                                                                      GlobalTransaction globalTransaction) {
+
+      Function<CacheTransaction, PendingTransaction> mapper = transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction) &&
+               transaction.containsLockOrBackupLock(key)) {
+            return new PendingTransaction(transaction, key);
+         }
+         return null;
+      };
+
+      return filterAndCollectTransactions(mapper);
+   }
+
+   private Collection<PendingTransaction> getTransactionWithAnyLockedKey(int transactionTopologyId,
+                                                                         Collection<Object> keys,
+                                                                         GlobalTransaction globalTransaction) {
+
+      Function<CacheTransaction, PendingTransaction> filter = transaction -> {
+         if (transaction.getTopologyId() < transactionTopologyId &&
+               !transaction.getGlobalTransaction().equals(globalTransaction)) {
+            Object key = transaction.findAnyLockedOrBackupLocked(keys.stream());
```

Comment:
you can't reuse a stream as you can't reuse iterators... I'll use the collection here but the cost is the same since I have to iterate over it...